runtime/drivers/olap.go

```diff
@@ -495,6 +500,128 @@ func (d Dialect) DateDiff(grain runtimev1.TimeGrain, t1, t2 time.Time) (string,
 	}
 }
 
+// SelectInlineResults returns a SQL query which inline results from the result set supplied.
+func (d Dialect) SelectInlineResults(result *Result) (string, []any, error) {
+	values := make([]any, len(result.Schema.Fields))
+	valuePtrs := make([]any, len(result.Schema.Fields))
+	for i := range values {
+		valuePtrs[i] = &values[i]
+	}
+
+	var dimVals []any
+
+	rows := 0
+	prefix := ""
+	suffix := ""
+	// creating inline query for all dialects in one loop, accumulating field exprs first and then creating the query can be more cleaner
+	for result.Next() {
+		if err := result.Scan(valuePtrs...); err != nil {
+			return "", nil, fmt.Errorf("select inline: failed to scan value: %w", err)
+		}
+		if d == DialectDruid {
+			// format - select * from (values (1, 2), (3, 4)) t(a, b)
+			if rows == 0 {
+				prefix = "SELECT * FROM (VALUES "
+				suffix = "t("
+			}
+			if rows > 0 {
+				prefix += ", "
+			}
+		} else {
+			// format - select 1 as a, 2 as b union all select 3 as a, 4 as b
+			if rows > 0 {
+				prefix += " UNION ALL "
+			}
+			prefix += "SELECT "
+		}
+
+		dimVals = append(dimVals, values[0])
+		for i, v := range values {
+			if d == DialectDruid {
+				if i == 0 {
+					prefix += "("
+				}
+				if rows == 0 {
+					suffix += d.EscapeIdentifier(result.Schema.Fields[i].Name)
+					if i != len(result.Schema.Fields)-1 {
+						suffix += ", "
+					}
+				}
+			}
+			if i > 0 {
+				prefix += ", "
+			}
+			ok, expr, err := d.GetValExpr(v, result.Schema.Fields[i].Type.Code)
+			if err != nil {
+				return "", nil, fmt.Errorf("select inline: failed to get value expression: %w", err)
+			}
+			if !ok {
+				return "", nil, fmt.Errorf("select inline: unsupported value type %q: %w", result.Schema.Fields[i].Type.Code, ErrOptimizationFailure)
+			}
+			if d == DialectDruid {
+				prefix += expr
+			} else {
+				prefix += fmt.Sprintf("%s AS %s", expr, d.EscapeIdentifier(result.Schema.Fields[i].Name))
+			}
+		}
+
+		if d == DialectDruid {
+			prefix += ")"
+			if rows == 0 {
+				suffix += ")"
+			}
+		}
+
+		rows++
+	}
+
+	if d == DialectDruid {
+		prefix += ") "
+	}
+
+	return prefix + suffix, dimVals, nil
+}
+
+func (d Dialect) GetValExpr(val any, typ runtimev1.Type_Code) (bool, string, error) {
+	if val == nil {
+		return true, "NULL", nil
+	}
+	switch typ {
+	case runtimev1.Type_CODE_STRING:
+		return true, fmt.Sprintf("'%v'", val), nil
```

Comment:
Is there a reason you didn't go with putting `?` and returning an array of positional args to bind to the query?

Reply:
changed to use positional args for CH and duckdb, druid does not support positional args for inline queries